contracts/hooks/FullRange.sol

```diff
@@ -0,0 +1,469 @@
+                FixedPoint128.Q128
+            )
+        );
+
+        position.feeGrowthInside0LastX128 = posInfo.feeGrowthInside0LastX128;
+        position.feeGrowthInside1LastX128 = posInfo.feeGrowthInside1LastX128;
+        // subtraction is safe because we checked positionLiquidity is gte liquidity
+        position.liquidity = uint128(positionLiquidity - liquidity);
+    }
+
+    // deploy ERC-20 contract
+    function beforeInitialize(address, IPoolManager.PoolKey calldata key, uint160) external override returns (bytes4) {
+        require(key.tickSpacing == 60, "Tick spacing must be default");
+
+        // deploy erc20 contract
+
+        bytes memory bytecode = type(UniswapV4ERC20).creationCode;
+        bytes32 salt = keccak256(abi.encodePacked(key.toId()));
+
+        address poolToken;
+        assembly {
+            poolToken := create2(0, add(bytecode, 32), mload(bytecode), salt)
+        }
+
+        poolToERC20[key.toId()] = poolToken;
+
+        return FullRange.beforeInitialize.selector;
+    }
+
+    function _rebalance(IPoolManager.PoolKey calldata key) internal {
+        console.log("we're rebalancing");
+        // retrieve all liquidity -- i think i should already have the liquidity kept track of
+        // uint128 hookLiquidity = poolManager.getLiquidity(key.toId(), address(this), MIN_TICK, MAX_TICK);
+
+        // position.liquidity = hookLiquidity;
+
+        HookPosition storage position = poolToHookPosition[key.toId()];
+
+        IPoolManager.ModifyPositionParams memory params = IPoolManager.ModifyPositionParams({
+            tickLower: MIN_TICK,
+            tickUpper: MAX_TICK,
+            liquidityDelta: -int256(int128(position.liquidity))
+        });
+
+        console.log("the entire hook liquidity");
+        console.log(position.liquidity);
+
+        BalanceDelta balanceDelta = hookModifyPosition(key, params);
+
+        (uint160 sqrtPriceX96,,,,,) = poolManager.getSlot0(key.toId());
+
+        uint128 liquidity = LiquidityAmounts.getLiquidityForAmounts(
+            sqrtPriceX96,
+            TickMath.getSqrtRatioAtTick(MIN_TICK),
+            TickMath.getSqrtRatioAtTick(MAX_TICK),
+            uint256(uint128(-balanceDelta.amount0())),
+            uint256(uint128(-balanceDelta.amount1()))
+        );
+
+        params = IPoolManager.ModifyPositionParams({
+            tickLower: MIN_TICK,
+            tickUpper: MAX_TICK,
+            liquidityDelta: int256(int128(liquidity))
+        });
+
+        // reinvest everything
+        hookModifyPosition(key, params);
+
+        // update position
+        Position.Info memory posInfo = poolManager.getPosition(key.toId(), address(this), MIN_TICK, MAX_TICK);
+
+        position.feeGrowthInside0LastX128 = posInfo.feeGrowthInside0LastX128;
+        position.feeGrowthInside1LastX128 = posInfo.feeGrowthInside1LastX128;
+        position.tokensOwed0 = 0;
+        position.tokensOwed1 = 0;
+    }
+
+    function beforeModifyPosition(
```

Comment:
nit, I'd stick these towards the top to follow the conventional organization of the rest of the hooks in the repo